### ISSUE TITLE
PVs becomes unreachable running IOs for sustained duration

### ISSUE THREAD

**Issue by maikotz** [openebs/openebs#176]
After a network interruption and a node reboot one of the PV is not reachable anymore. 

The logoutput of the ctrl pod shows:
```
time="2017-07-22T10:43:17Z" level=info msg="Connecting to remote: 10.1.0.55:9502" 
time="2017-07-22T10:43:17Z" level=error msg="Error in request: json: cannot unmarshal number 1.5484855e+07 into Go value of type int64" 
10.1.0.55 - - [22/Jul/2017:10:43:17 +0000] "POST /v1/volumes/cHZjLTRhM2Q0NzYzLTY3YTYtMTFlNy05ODIxLTI0NmU5NjFjMzZhOA==?action=start HTTP/1.1" 500 275
10.1.7.94 - - [22/Jul/2017:10:43:17 +0000] "POST /v1/register HTTP/1.1" 0 0
10.1.7.94 - - [22/Jul/2017:10:43:19 +0000] "POST /v1/register HTTP/1.1" 0 0
10.1.0.55 - - [22/Jul/2017:10:43:19 +0000] "POST /v1/register HTTP/1.1" 0 0
10.1.7.94 - - [22/Jul/2017:10:43:24 +0000] "POST /v1/register HTTP/1.1" 0 0
time="2017-07-22T10:44:19Z" level=info msg="Connecting to remote: 10.1.0.55:9502" 
time="2017-07-22T10:44:19Z" level=error msg="Error in request: json: cannot unmarshal number 1.5484855e+07 into Go value of type int64" 
10.1.0.55 - - [22/Jul/2017:10:44:19 +0000] "POST /v1/volumes/cHZjLTRhM2Q0NzYzLTY3YTYtMTFlNy05ODIxLTI0NmU5NjFjMzZhOA==?action=start HTTP/1.1" 500 275
```

Restarting the pods didn't help. And even with apparently one of the two replica pods having  an established connection to the ctrl pod  the connection can not be established.

```
time="2017-07-22T09:51:19Z" level=info msg="Listening on sync 10.1.0.55:9504" 
10.1.0.56 - - [22/Jul/2017:09:52:59 +0000] "POST /v1/replicas/1?action=start HTTP/1.1" 200 904
time="2017-07-22T09:52:59Z" level=info msg="New connection from: 10.1.0.56:58198" 
time="2017-07-22T09:52:59Z" level=info msg="Opening volume /openebs, size 85899345920/512" 
10.1.0.56 - - [22/Jul/2017:09:52:59 +0000] "POST /v1/replicas/1?action=open HTTP/1.1" 200 1744
10.1.0.56 - - [22/Jul/2017:09:53:01 +0000] "POST /v1/replicas/1?action=start HTTP/1.1" 200 1744
```

I have a couple of pods in the testing environment that this morning went down and would appreciate any ideas

**Comment by gila** [comment]
Thanks for the report, we will look into it.

**Comment by kmova** [comment]
@maikotz 
Can you please send the logs of each of the replica pods? 
The K8s events on the controller and replica also will help. (kubectl describe ..)

**Comment by maikotz** [comment]
Hello @kmova 

the logs can be found on gist: [Gist](https://gist.github.com/maikotz/98ff8f80fba73d620e8385e6fdb5fa88)

It seems that the iscsid is not listening. The connections to all ports on the replication containers seem to work (at least nmap says they're open ports) but if I try to connect from another container to the service IP with port 3260 I don't get a connection. Same if I try the pod IP.

**Comment by kmova** [comment]
Thanks @maikotz. The controller logs show that the 3260 isn't opened yet. Will look into that error 500 and get back with a new build.

**Comment by maikotz** [comment]
Ok I dug a bit more and saw that the value which apparently can't be transformed is the revision counter.

Having `  "revisioncounter": 15484855,`in the answer of register I get the message:

`time="2017-07-24T10:43:04Z" level=error msg="Error in request: json: cannot unmarshal number 1.5484855e+07 into Go value of type int64" ` 

That number matches: `/var/openebs/pvc-4a3d4763-67a6-11e7-9821-246e961c36a8/openebs/revision.counter` on one host machine. On the second hostmachine its higher than that:
`sudo cat /var/openebs/pvc-4a3d4763-67a6-11e7-9821-246e961c36a8/openebs/revision.counter
16992745`

All the pods that went down are either elasticsearch data pods or mariadb pods with high revision counters

Other pods (mailer pods, persistent storage for websites, etc) with smaller revision numbers and a lot less traffic are not affected and still functional.  

Could this indicate a problem with higher io traffic and replication ?

**Comment by maikotz** [comment]
Edit: [GIST to tcpdump](https://gist.github.com/maikotz/c2ec1c0807b14946592f111263f88129)

This is the connection between ctrl and rep where the error occurs. 

is it "normal" that the links and actions are both "null" ?

`{"links":null,"actions":null,"replicas":["tcp://10.1.7.172:9502"]}`

**Comment by payes** [comment]
@maikotz 
Thanks for helping debug the issue.
As stated in this link:
[https://github.com/golang/go/issues/5562](url)
the problem is with the large values of revision counter.
We are not able to unmarshal the revision counter at the controller sent by replica while registering themselves at the controller.
We will fix this issue and create a new docker image which can be used.

**Comment by maikotz** [comment]
awesome, thanks for your reply @payes

**Comment by kmova** [comment]
@payes  can you check for protobuf/flatbuf marshalling/unmarshalling to fix this issue? keep us posted on the findings.

**Comment by AmitKumarDas** [comment]
@payes Check below for reference of `protobuf in Go`

ref- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go & search for int64. It seems the marshalling/un-marshalling are controlled via protobuf & its datatypes here.

**Comment by kmova** [comment]
The issue was resolved in 0.4.0 with https://github.com/openebs/jiva/commit/6c914d62e7f3755f826db88e3d82fb17e9b2b4c2
